File: 2021/day06.cpp
#include <vector>

#include "../common/puzzle.hpp"

using namespace std;

class Solution : public Puzzle {
   public:
    Solution(std::string inputFileName = "inputs/2021/6.txt")
        : Puzzle(inputFileName) {}
    std::string part1() {
        int arr[9] = {0};
        for (int i = 0; i < 300; i++) {
            int a;
            inputFile >> a;
            inputFile.ignore(1);
            arr[a]++;
        }
        for (int i = 0; i < 80; i++) {
            int tmp = arr[0];
            for (int j = 0; j < 8; j++) arr[j] = arr[j + 1];
            arr[6] += tmp;
            arr[8] = tmp;
        }
        int sum = 0;
        for (int i = 0; i < 9; i++) sum += arr[i];
        cout << sum << "\n";
        return "TODO";
    }
    std::string part2() {
        long int arr[9] = {0};
        for (int i = 0; i < 300; i++) {
            int a;
            inputFile >> a;
            inputFile.ignore(1);
            arr[a]++;
        }
        for (int i = 0; i < 256; i++) {
            long int tmp = arr[0];
            for (int j = 0; j < 8; j++) arr[j] = arr[j + 1];
            arr[6] += tmp;
            arr[8] = tmp;
        }
        long int sum = 0;
        for (int i = 0; i < 9; i++) sum += arr[i];
        cout << sum << "\n";
        return "TODO";
    }
};

int main() {
    Solution s = Solution();
    s.solve();
}

File: year2016/day06.java
package year2016;

import java.util.Scanner;
import java.util.HashMap;
import java.util.Map;

import common.puzzle;

public class day06 extends puzzle {
    public day06() {
        super("./inputs/2016/6.txt");
    }

    public static void main(String[] args) {
        new day06().solve();
    }

    protected String part1() {
        Scanner reader = getScanner();
        int[][] counter = new int[8][26];
        while (reader.hasNextLine()) {
            char[] line = reader.nextLine().toCharArray();
            for (int i = 0; i < line.length; i++)
                counter[i][line[i] - 'a']++;
        }
        String result = "";
        for (int[] is : counter) {
            int maxIndex = 0;
            for (int i = 0; i < is.length; i++)
                if (is[maxIndex] < is[i])
                    maxIndex = i;
            result += (char) (maxIndex + 'a');
        }
        return result;
    }

    protected String part2() {
        Scanner reader = getScanner();
        int[][] counter = new int[8][26];
        while (reader.hasNextLine()) {
            char[] line = reader.nextLine().toCharArray();
            for (int i = 0; i < line.length; i++)
                counter[i][line[i] - 'a']++;
        }
        String result = "";
        for (int[] is : counter) {
            int minIndex = 0;
            for (int i = 0; i < is.length; i++)
                if (is[minIndex] > is[i])
                    minIndex = i;
            result += (char) (minIndex + 'a');
        }
        return result;
    }
}